File: crates/ans-protocol/src/derive.rs
use arch_program::pubkey::Pubkey;
use sha2::{Digest, Sha256};

use crate::state::{ArchAddress, RecordType};

const CONFIG_SEED: &[u8] = b"ans:config:v1";
const NAME_SEED: &[u8] = b"ans:name:v1";
const RECORD_SEED: &[u8] = b"ans:record:v1";
const REVERSE_SEED: &[u8] = b"ans:reverse:v1";
const NAMESPACE_DOMAIN: &[u8] = b"arch-name-service:namespace:v1\0";
const RECORD_KEY_DOMAIN: &[u8] = b"arch-name-service:record-key:v1\0";

pub fn namespace_hash(namespace: &str) -> [u8; 32] {
    let mut hasher = Sha256::new();
    hasher.update(NAMESPACE_DOMAIN);
    hasher.update(namespace.as_bytes());
    hasher.finalize().into()
}

/// Domain-separated hash of a Text record key for PDA seeds.
pub fn record_key_hash(key: &str) -> [u8; 32] {
    let mut hasher = Sha256::new();
    hasher.update(RECORD_KEY_DOMAIN);
    hasher.update(key.as_bytes());
    hasher.finalize().into()
}

pub fn derive_config_address(
    program_id: ArchAddress,
    network_id: u32,
    namespace: &str,
) -> ArchAddress {
    derive(
        program_id,
        &[
            CONFIG_SEED,
            &network_id.to_le_bytes(),
            &namespace_hash(namespace),
        ],
    )
}

pub fn derive_name_address(
    program_id: ArchAddress,
    namespace: &str,
    name_hash: [u8; 32],
) -> ArchAddress {
    derive(
        program_id,
        &[NAME_SEED, &namespace_hash(namespace), &name_hash],
    )
}

/// Derives a typed record PDA (`ArchOwner` / `BitcoinTaproot` / `TokenAta`).
/// For `Text` records use [`derive_text_record_address`]. Typed PDAs stay
/// byte-identical to program version 1 so live names are not orphaned.
pub fn derive_record_address(
    program_id: ArchAddress,
    namespace: &str,
    name_hash: [u8; 32],
    record_type: RecordType,
) -> ArchAddress {
    debug_assert!(
        record_type != RecordType::Text,
        "Text records require derive_text_record_address"
    );
    derive(
        program_id,
        &[
            RECORD_SEED,
            &namespace_hash(namespace),
            &name_hash,
            &[record_type as u8],
        ],
    )
}

/// Derives a Text record PDA for `(name_hash, key)`. Existing typed-record
/// addresses are unchanged; only the new `Text` discriminant + key hash seed
/// are added.
pub fn derive_text_record_address(
    program_id: ArchAddress,
    namespace: &str,
    name_hash: [u8; 32],
    key: &str,
) -> ArchAddress {
    let key_hash = record_key_hash(key);
    derive(
        program_id,
        &[
            RECORD_SEED,
            &namespace_hash(namespace),
            &name_hash,
            &[RecordType::Text as u8],
            &key_hash,
        ],
    )
}

pub fn derive_record_address_for_value(
    program_id: ArchAddress,
    namespace: &str,
    name_hash: [u8; 32],
    record_type: RecordType,
    text_key: Option<&str>,
) -> ArchAddress {
    match record_type {
        RecordType::Text => derive_text_record_address(
            program_id,
            namespace,
            name_hash,
            text_key.expect("Text record requires a key"),
        ),
        _ => derive_record_address(program_id, namespace, name_hash, record_type),
    }
}

pub fn derive_reverse_address(
    program_id: ArchAddress,
    namespace: &str,
    owner: ArchAddress,
) -> ArchAddress {
    derive(
        program_id,
        &[REVERSE_SEED, &namespace_hash(namespace), &owner],
    )
}

fn derive(program_id: ArchAddress, seeds: &[&[u8]]) -> ArchAddress {
    let program_id = Pubkey::new_from_array(program_id);
    Pubkey::find_program_address(seeds, &program_id)
        .0
        .serialize()
}

File: packages/ans-sdk/src/derive.ts
import { PubkeyUtil } from "@saturnbtcio/arch-sdk";

import {
  CONFIG_SEED,
  LISTING_SEED,
  NAME_SEED,
  OFFER_SEED,
  RECORD_SEED,
  REVERSE_SEED,
} from "./constants.js";
import { namespaceHash, recordKeyHash } from "./hash.js";
import type { ArchAddress, RecordType } from "./types.js";

function encodeU32LE(value: number): Uint8Array {
  const out = new Uint8Array(4);
  new DataView(out.buffer).setUint32(0, value >>> 0, true);
  return out;
}

function recordTypeByte(recordType: RecordType): number {
  switch (recordType) {
    case "ArchOwner":
      return 0;
    case "BitcoinTaproot":
      return 1;
    case "TokenAta":
      return 2;
    case "Text":
      return 3;
  }
}

function asBytes(value: Uint8Array): Uint8Array {
  // Copy into a fresh Uint8Array so Node Buffers / cross-realm views
  // satisfy @noble/hashes' strict byte checks inside arch-sdk PDA helpers.
  return Uint8Array.from(value);
}

function derive(programId: ArchAddress, seeds: Uint8Array[]): ArchAddress {
  return PubkeyUtil.findProgramAddress(
    seeds.map(asBytes),
    asBytes(programId),
  )[0];
}

export function deriveConfigAddress(
  programId: ArchAddress,
  networkId: number,
  namespace: string,
): ArchAddress {
  return derive(programId, [
    CONFIG_SEED,
    encodeU32LE(networkId),
    namespaceHash(namespace),
  ]);
}

export function deriveNameAddress(
  programId: ArchAddress,
  namespace: string,
  nameHashBytes: Uint8Array,
): ArchAddress {
  return derive(programId, [NAME_SEED, namespaceHash(namespace), nameHashBytes]);
}

/** Typed records only (ArchOwner / BitcoinTaproot / TokenAta). */
export function deriveRecordAddress(
  programId: ArchAddress,
  namespace: string,
  nameHashBytes: Uint8Array,
  recordType: Exclude<RecordType, "Text">,
): ArchAddress {
  return derive(programId, [
    RECORD_SEED,
    namespaceHash(namespace),
    nameHashBytes,
    Uint8Array.of(recordTypeByte(recordType)),
  ]);
}

export function deriveTextRecordAddress(
  programId: ArchAddress,
  namespace: string,
  nameHashBytes: Uint8Array,
  key: string,
): ArchAddress {
  return derive(programId, [
    RECORD_SEED,
    namespaceHash(namespace),
    nameHashBytes,
    Uint8Array.of(recordTypeByte("Text")),
    recordKeyHash(key),
  ]);
}

export function deriveRecordAddressFor(
  programId: ArchAddress,
  namespace: string,
  nameHashBytes: Uint8Array,
  recordType: RecordType,
  textKey?: string,
): ArchAddress {
  if (recordType === "Text") {
    if (!textKey) {
      throw new Error("Text records require a key");
    }
    return deriveTextRecordAddress(programId, namespace, nameHashBytes, textKey);
  }
  return deriveRecordAddress(programId, namespace, nameHashBytes, recordType);
}

export function deriveReverseAddress(
  programId: ArchAddress,
  namespace: string,
  owner: ArchAddress,
): ArchAddress {
  return derive(programId, [REVERSE_SEED, namespaceHash(namespace), owner]);
}

export function deriveListingAddress(
  programId: ArchAddress,
  namespace: string,
  nameHashBytes: Uint8Array,
): ArchAddress {
  return derive(programId, [LISTING_SEED, namespaceHash(namespace), nameHashBytes]);
}

export function deriveOfferAddress(
  programId: ArchAddress,
  namespace: string,
  nameHashBytes: Uint8Array,
  buyer: ArchAddress,
): ArchAddress {
  return derive(programId, [
    OFFER_SEED,
    namespaceHash(namespace),
    nameHashBytes,
    buyer,
  ]);
}

export function deriveTokenAta(
  owner: ArchAddress,
  tokenId: ArchAddress,
  tokenProgramId: ArchAddress,
  associatedTokenProgramId: ArchAddress,
): ArchAddress {
  return PubkeyUtil.findProgramAddress(
    [owner, tokenProgramId, tokenId],
    associatedTokenProgramId,
  )[0];
}
